CIMPLR - Articial data experiment
========================================================

```{r}

### Load libraries
library(Biobase)
library(GenomicFeatures)
library(rjson)
library(multicore)
library(rtracklayer)
library(raster)
library(RMySQL)


source('R/AllClasses.R')
source('R/AllGenerics.R')
source('R/KSEDistribution.R')

source('R/core.R')
source('R/cimplr.R')
source('R/artificial_db.R')

```




```{r}

conn <- dbConnect(drv=MySQL(max.con=100), host='medoid', dbname='im_simulation_db', user='igkc', password='igkc')

artificial_iset_ids <- getArtificialIsetIds(conn)

actual <- getArtificialCIS(conn)
iset <- getArtificialIset(conn, artificial_iset_ids[500])

co  <- cimplr(
  insertions        = iset,
  
  biasmap           = 'biasmaps_chr10/bsgenomes-mm9-TA-stepsize100-D4',
  biasmap.type      = 'count',

  scales            = log.seq(500, 500000, 50),
  
  alpha             = 0.05,
  p.adjust.method   = 'fdr',

	n.cores           = c(scales=10, chromosomes=1),
	genes.file        = NULL
)


caller <- c(
	list(
		method='cimplr'
	),
	co$input[c('biasmap', 'biasmap.type', 'scales', 'alpha', 'p.adjust.method')]
)

caller$scales <- .formatScales(caller$scales)

sql <- paste("DELETE FROM `evaluation` WHERE caller = '", toJSON(caller), "'", sep='')
dbGetQuery(conn, sql)
dbDisconnect(conn)
```




```{r}


co <- cimplr.biasmap(co)
co <- cimplr.convolve(co)




```




```{r}

artificial_iset_ids <- sample(artificial_iset_ids)
artificial_iset_ids <- 1000

mcresult <- mclapply (artificial_iset_ids, mc.cores=10, function(artificial_iset_id) {
  
	progress <- paste(which(artificial_iset_ids == artificial_iset_id), '/', length(artificial_iset_ids), sep='')

	message(progress, ' - cimplr (iset_id = ', artificial_iset_id, ')')

	
	
	# load artificial insertion set
	conn <- dbConnect(drv=MySQL(max.con=100), host='medoid', dbname='im_simulation_db', user='igkc', password='igkc')
	iset <- getArtificialIset(conn, artificial_iset_id)



	co <- cimplr.replace_insertions(co, iset)
	co <- cimplr.convolve(co)
  co <- cimplr.threshold(co)
	co <- cimplr.call(co)



	message(progress, ' - evaluate (iset_id = ', artificial_iset_id, ')')

	predicted_cis <- co$output$collapsed.cises


	if (is.null(predicted_cis)) {
		predicted <- GRanges()
	} else {

		predicted <- GRanges(
			seqnames = Rle(predicted_cis$chromosome),
			ranges = IRanges(
				start = predicted_cis$start,
				end = predicted_cis$end
			)
		)
	}


	writeEvaluation(conn, actual, predicted, artificial_iset_id, caller)

	dbDisconnect(conn)

	if ((artificial_iset_id %% 100) == 0) {
		message(progress, ' - export (iset_id = ', artificial_iset_id, ')')
		art.output.dir  <- paste('../cimplr_projects/artificial/iset', artificial_iset_id, sep='')
		dir.create(art.output.dir, showWarnings=FALSE, recursive=TRUE)

		export.cimplr(co, output.dir = art.output.dir)
	}

})

message('')
message('end loop for ', length(artificial_iset_ids), " iset's")
message('')




```