CIMPLR - Articial data experiment
========================================================

```{r}

### Load libraries
library(Biobase)
library(GenomicFeatures)
library(rjson)
library(multicore)
library(rtracklayer)
library(raster)
library(RMySQL)


source('lib/AllClasses.R')
source('lib/AllGenerics.R')
source('lib/InsertionSet.R')
source('lib/KSEDistribution.R')

source('lib/core.R')
source('lib/cimplr.R')
source('lib/artificial_db.R')



# DELETE FROM `evaluation` WHERE `evaluation_id` > 7803;
# DELETE FROM `evaluation_artificial_cis` WHERE `evaluation_id` > 7803;
# DELETE FROM `evaluation_predicted_cis` WHERE `evaluation_id` > 7803;


#p.adjust.method <- 'n.peaks'




conn <- dbConnect(drv=MySQL(max.con=100), host='medoid', dbname='im_simulation_db', user='igkc', password='igkc')

artificial_iset_ids <- getArtificialIsetIds(conn)

actual <- getArtificialCIS(conn)
iset <- getArtificialIset(conn, artificial_iset_ids[500])

co  <- cimplr(
  insertions        = iset,
	reference         = 'mm9',
	pattern           = 'TA',
	scales            = log.seq(5000, 500000, 100)[seq(1,50,5)],
	pkse.method       = 'count',
	alpha.level       = 0.05,
	p.adjust.method   = 'bonferroni',
	p.adjust.n.method = 'n.x',
	mtf               = 1,
	test.chromosomes.seperately = TRUE,
	n.cores           = c(scales=10, chromosomes=1),
	genes.file        = NULL
)


caller <- c(
	list(
		method='cimplr'
	),
	co$input[c('reference', 'pattern', 'scales', 'pkse.method', 'alpha.level', 'p.adjust.method', 'p.adjust.n.method', 'mtf')]
)

caller$scales <- .formatScales(caller$scales)

sql <- paste("DELETE FROM `evaluation` WHERE caller = '", toJSON(caller), "'", sep='')
dbGetQuery(conn, sql)
dbDisconnect(conn)


co <- cimplr.reference(co)
#co$input$n.cores <- c(scales=1, chromosomes=1)
co <- cimplr.convolve(co) # to calulate threshold

#stop()

message('')
message('starting loop for ', length(artificial_iset_ids), " iset's")
message('')


artificial_iset_ids <- sample(artificial_iset_ids)


#$ 1000:Class 'try-error'  atomic [1:1] Error in unique(do.call("rbind", lapply(1:nrow(maxXY), function(i) { : 
#  error in evaluating the argument 'x' in selecting a method for function 'unique': Error: sum(idx) == 1 is not TRUE
#
artificial_iset_ids <- 1000

#artificial_iset_ids <- 1085
source('scripts/cimplr_artificial_loop.R')
mcresult <- mclapply (artificial_iset_ids, mc.cores=10, function(artificial_iset_id) {
  
	progress <- paste(which(artificial_iset_ids == artificial_iset_id), '/', length(artificial_iset_ids), sep='')

	message(progress, ' - cimplr (iset_id = ', artificial_iset_id, ')')

	
	
	# load artificial insertion set
	conn <- dbConnect(drv=MySQL(max.con=100), host='medoid', dbname='im_simulation_db', user='igkc', password='igkc')
	iset <- getArtificialIset(conn, artificial_iset_id)



	co <- cimplr.replace_insertions(co, iset)
	co <- cimplr.convolve(co)
  co <- cimplr.threshold(co)
	co <- cimplr.call(co)



	message(progress, ' - evaluate (iset_id = ', artificial_iset_id, ')')

	predicted_cis <- co$output$collapsed.cises


	if (is.null(predicted_cis)) {
		predicted <- GRanges()
	} else {

		predicted <- GRanges(
			seqnames = Rle(predicted_cis$chromosome),
			ranges = IRanges(
				start = predicted_cis$start,
				end = predicted_cis$end
			)
		)
	}


	writeEvaluation(conn, actual, predicted, artificial_iset_id, caller)

	dbDisconnect(conn)

	if ((artificial_iset_id %% 100) == 0) {
		message(progress, ' - export (iset_id = ', artificial_iset_id, ')')
		art.output.dir  <- paste('../cimplr_projects/artificial/iset', artificial_iset_id, sep='')
		dir.create(art.output.dir, showWarnings=FALSE, recursive=TRUE)

		export.cimplr(co, output.dir = art.output.dir)
	}

})

message('')
message('end loop for ', length(artificial_iset_ids), " iset's")
message('')




```