CIMPLR
========================================================

A package to call Common Insertion Sites using biasmaps without permutation testing.

In preparation for running the actual CIMPLR commands, we start by preparing an R environment containing CIMPLR functions and insertion data.

First set your working directory to the root folder of the package. Here we do that by setting the 'Knit' options to create this Vignette; use `setwd('../')` at the console.

```{r opts_knit}
opts_knit$set(root.dir='../')
```

Load the necessary libraries and sources of the package.

```{r libraries}

library(Biobase)
library(GenomicFeatures)
library(rjson)
library(multicore)
library(rtracklayer)
library(raster)

source('R/AllClasses.R')
source('R/AllGenerics.R')
source('R/KSEDistribution.R')
source('R/core.R')
source('R/cimplr.R')
source('R/cimplr-calc.R')
source('R/cimplr-export.R')
source('R/biasmaps.R')
```

Load your data as a `data.frame` containing at least a `chr` and `location` column.

```{r load_an_iset}
load('data/test_iset.rda')
#iset <- read.table(file='./data//WONG_NATURE-GENETICS_2014_INSERTIONS_MM10_FILTERED_BY_CONTROLS.cimpl', sep="\t", header=TRUE)

head(iset)
```



Now we can start with CIMPLR.


Step 1: Initialise a CIMPLR object
-------------------------

We construct a CIMPLR object based on the insertion data and the parameters for the CIMPLR run.

```{r initialize_a_cimplr_object}
co  <- cimplr(
  insertions        = iset,
  exclude.chromosomes = c('chrY'),
  
  biasmap           = 'biasmaps_chr10/bsgenomes-mm9-TA-stepsize100-D4',
#  biasmap           = 'biasmaps/bsgenomes-mm9-TA-stepsize1000-D4',
  biasmap.type      = 'count',
  
  scales            = log.seq(500, 500000, 50)[seq(21, 40, 5)],
  alpha             = 0.05,
  p.adjust.method   = 'fdr',
  
	n.cores           = 4,
	genes.file        = NULL
)
```


Step 2: Load a bias map
-------------------------

As a bias map, we load the TA count files under the './biasmaps' folder. This bias map is simply the TA site background, but more sophisticated bias maps can be supplied here (link to CreateBiasMap Vignette).

```{r set_reference}
co <- cimplr.biasmap(co)
```


```{r plot_bg, fig.width=10, fig.height=6}
plot.chrscale(co, 'chr10', 34332, 'bg')
```


Step 3: Gaussian Kernel Convolution
-------------------------

With the insertion data and the bias map loaded and pre-processed within the CIMPLR object, we can now start the kernel convolution, resulting in Kernel Smoothed Estimates (KSEs) for all scales.
```{r convolve}
co <- cimplr.convolve(co)
```

```{r plot_kse, fig.width=10, fig.height=6}
plot.chrscale(co, 'chr10', 34332, 'kse')
```

```{r plot_p, fig.width=10, fig.height=6}
plot.chrscale(co, 'chr10', 34332, 'p')
```


Step 4: Calculate the threshold and the kse/ratio's
-------------------------

A seperate step calculates the threshold for the KSE at each scale. This is done using the analytical KSE distribution at a controlled FDR and taking into account multiple scales (link to KSEdistribution Vignette).
```{r calculcate_ratios}
co <- cimplr.ratios(co)
```

```{r plot_ratio, fig.width=10, fig.height=6}
plot.chrscale(co, 'chr10', 34332, 'ratio')
```



Step 5: Call CISes
-------------------------

With the thresholds in place, the last step is to do the CIS calling. This returns two lists of CISes: (1) all CIS for each scale, and (2) the cross scale CIS using the 'cis-cloud-global-maximum' procedure (link to CrossScaleCIScalling Vignette)

```{r call_cis}
co <- cimplr.call(co)

head(co$output$all.cises)
head(co$output$collapsed.cises)
```



Step 6: Output figures and cis list. The export function generates:
- all_cises.csv
- collapsed_cises.csv (according to the 'cis-cloud-global-maximum' procedure)
- a scale space figure per chromosome, focussing on the collapsed CIS regions
- figures and CIS lists per scale (include.scales=TRUE)


```{r export}
#cimplr.annotate()
#export.cimplr(co, './test_output/', include.scale=TRUE)
```



