<!DOCTYPE html>
<!-- saved from url=(0014)about:internet -->
<html>
<head>
<meta http-equiv="Content-Type" content="text/html; charset=utf-8"/>
<meta http-equiv="x-ua-compatible" content="IE=9" >

<title>Using CIMPLR to call Common Insertion Sites using bias maps without permutation testing</title>

<style type="text/css">
body, td {
   font-family: sans-serif;
   background-color: white;
   font-size: 12px;
   margin: 8px;
}

tt, code, pre {
   font-family: 'DejaVu Sans Mono', 'Droid Sans Mono', 'Lucida Console', Consolas, Monaco, monospace;
}

h1 { 
   font-size:2.2em; 
}

h2 { 
   font-size:1.8em; 
}

h3 { 
   font-size:1.4em; 
}

h4 { 
   font-size:1.0em; 
}

h5 { 
   font-size:0.9em; 
}

h6 { 
   font-size:0.8em; 
}

a:visited {
   color: rgb(50%, 0%, 50%);
}

pre {	
   margin-top: 0;
   max-width: 95%;
   border: 1px solid #ccc;
   white-space: pre-wrap;
}

pre code {
   display: block; padding: 0.5em;
}

code.r, code.cpp {
   background-color: #F8F8F8;
}

table, td, th {
  border: none;
}

blockquote {
   color:#666666;
   margin:0;
   padding-left: 1em;
   border-left: 0.5em #EEE solid;
}

hr {
   height: 0px;
   border-bottom: none;
   border-top-width: thin;
   border-top-style: dotted;
   border-top-color: #999999;
}

@media print {
   * { 
      background: transparent !important; 
      color: black !important; 
      filter:none !important; 
      -ms-filter: none !important; 
   }

   body { 
      font-size:12pt; 
      max-width:100%; 
   }
       
   a, a:visited { 
      text-decoration: underline; 
   }

   hr { 
      visibility: hidden;
      page-break-before: always;
   }

   pre, blockquote { 
      padding-right: 1em; 
      page-break-inside: avoid; 
   }

   tr, img { 
      page-break-inside: avoid; 
   }

   img { 
      max-width: 100% !important; 
   }

   @page :left { 
      margin: 15mm 20mm 15mm 10mm; 
   }
     
   @page :right { 
      margin: 15mm 10mm 15mm 20mm; 
   }

   p, h2, h3 { 
      orphans: 3; widows: 3; 
   }

   h2, h3 { 
      page-break-after: avoid; 
   }
}

</style>

<!-- Styles for R syntax highlighter -->
<style type="text/css">
   pre .operator,
   pre .paren {
     color: rgb(104, 118, 135)
   }

   pre .literal {
     color: rgb(88, 72, 246)
   }

   pre .number {
     color: rgb(0, 0, 205);
   }

   pre .comment {
     color: rgb(76, 136, 107);
   }

   pre .keyword {
     color: rgb(0, 0, 255);
   }

   pre .identifier {
     color: rgb(0, 0, 0);
   }

   pre .string {
     color: rgb(3, 106, 7);
   }
</style>

<!-- R syntax highlighter -->
<script type="text/javascript">
var hljs=new function(){function m(p){return p.replace(/&/gm,"&amp;").replace(/</gm,"&lt;")}function f(r,q,p){return RegExp(q,"m"+(r.cI?"i":"")+(p?"g":""))}function b(r){for(var p=0;p<r.childNodes.length;p++){var q=r.childNodes[p];if(q.nodeName=="CODE"){return q}if(!(q.nodeType==3&&q.nodeValue.match(/\s+/))){break}}}function h(t,s){var p="";for(var r=0;r<t.childNodes.length;r++){if(t.childNodes[r].nodeType==3){var q=t.childNodes[r].nodeValue;if(s){q=q.replace(/\n/g,"")}p+=q}else{if(t.childNodes[r].nodeName=="BR"){p+="\n"}else{p+=h(t.childNodes[r])}}}if(/MSIE [678]/.test(navigator.userAgent)){p=p.replace(/\r/g,"\n")}return p}function a(s){var r=s.className.split(/\s+/);r=r.concat(s.parentNode.className.split(/\s+/));for(var q=0;q<r.length;q++){var p=r[q].replace(/^language-/,"");if(e[p]){return p}}}function c(q){var p=[];(function(s,t){for(var r=0;r<s.childNodes.length;r++){if(s.childNodes[r].nodeType==3){t+=s.childNodes[r].nodeValue.length}else{if(s.childNodes[r].nodeName=="BR"){t+=1}else{if(s.childNodes[r].nodeType==1){p.push({event:"start",offset:t,node:s.childNodes[r]});t=arguments.callee(s.childNodes[r],t);p.push({event:"stop",offset:t,node:s.childNodes[r]})}}}}return t})(q,0);return p}function k(y,w,x){var q=0;var z="";var s=[];function u(){if(y.length&&w.length){if(y[0].offset!=w[0].offset){return(y[0].offset<w[0].offset)?y:w}else{return w[0].event=="start"?y:w}}else{return y.length?y:w}}function t(D){var A="<"+D.nodeName.toLowerCase();for(var B=0;B<D.attributes.length;B++){var C=D.attributes[B];A+=" "+C.nodeName.toLowerCase();if(C.value!==undefined&&C.value!==false&&C.value!==null){A+='="'+m(C.value)+'"'}}return A+">"}while(y.length||w.length){var v=u().splice(0,1)[0];z+=m(x.substr(q,v.offset-q));q=v.offset;if(v.event=="start"){z+=t(v.node);s.push(v.node)}else{if(v.event=="stop"){var p,r=s.length;do{r--;p=s[r];z+=("</"+p.nodeName.toLowerCase()+">")}while(p!=v.node);s.splice(r,1);while(r<s.length){z+=t(s[r]);r++}}}}return z+m(x.substr(q))}function j(){function q(x,y,v){if(x.compiled){return}var u;var s=[];if(x.k){x.lR=f(y,x.l||hljs.IR,true);for(var w in x.k){if(!x.k.hasOwnProperty(w)){continue}if(x.k[w] instanceof Object){u=x.k[w]}else{u=x.k;w="keyword"}for(var r in u){if(!u.hasOwnProperty(r)){continue}x.k[r]=[w,u[r]];s.push(r)}}}if(!v){if(x.bWK){x.b="\\b("+s.join("|")+")\\s"}x.bR=f(y,x.b?x.b:"\\B|\\b");if(!x.e&&!x.eW){x.e="\\B|\\b"}if(x.e){x.eR=f(y,x.e)}}if(x.i){x.iR=f(y,x.i)}if(x.r===undefined){x.r=1}if(!x.c){x.c=[]}x.compiled=true;for(var t=0;t<x.c.length;t++){if(x.c[t]=="self"){x.c[t]=x}q(x.c[t],y,false)}if(x.starts){q(x.starts,y,false)}}for(var p in e){if(!e.hasOwnProperty(p)){continue}q(e[p].dM,e[p],true)}}function d(B,C){if(!j.called){j();j.called=true}function q(r,M){for(var L=0;L<M.c.length;L++){if((M.c[L].bR.exec(r)||[null])[0]==r){return M.c[L]}}}function v(L,r){if(D[L].e&&D[L].eR.test(r)){return 1}if(D[L].eW){var M=v(L-1,r);return M?M+1:0}return 0}function w(r,L){return L.i&&L.iR.test(r)}function K(N,O){var M=[];for(var L=0;L<N.c.length;L++){M.push(N.c[L].b)}var r=D.length-1;do{if(D[r].e){M.push(D[r].e)}r--}while(D[r+1].eW);if(N.i){M.push(N.i)}return f(O,M.join("|"),true)}function p(M,L){var N=D[D.length-1];if(!N.t){N.t=K(N,E)}N.t.lastIndex=L;var r=N.t.exec(M);return r?[M.substr(L,r.index-L),r[0],false]:[M.substr(L),"",true]}function z(N,r){var L=E.cI?r[0].toLowerCase():r[0];var M=N.k[L];if(M&&M instanceof Array){return M}return false}function F(L,P){L=m(L);if(!P.k){return L}var r="";var O=0;P.lR.lastIndex=0;var M=P.lR.exec(L);while(M){r+=L.substr(O,M.index-O);var N=z(P,M);if(N){x+=N[1];r+='<span class="'+N[0]+'">'+M[0]+"</span>"}else{r+=M[0]}O=P.lR.lastIndex;M=P.lR.exec(L)}return r+L.substr(O,L.length-O)}function J(L,M){if(M.sL&&e[M.sL]){var r=d(M.sL,L);x+=r.keyword_count;return r.value}else{return F(L,M)}}function I(M,r){var L=M.cN?'<span class="'+M.cN+'">':"";if(M.rB){y+=L;M.buffer=""}else{if(M.eB){y+=m(r)+L;M.buffer=""}else{y+=L;M.buffer=r}}D.push(M);A+=M.r}function G(N,M,Q){var R=D[D.length-1];if(Q){y+=J(R.buffer+N,R);return false}var P=q(M,R);if(P){y+=J(R.buffer+N,R);I(P,M);return P.rB}var L=v(D.length-1,M);if(L){var O=R.cN?"</span>":"";if(R.rE){y+=J(R.buffer+N,R)+O}else{if(R.eE){y+=J(R.buffer+N,R)+O+m(M)}else{y+=J(R.buffer+N+M,R)+O}}while(L>1){O=D[D.length-2].cN?"</span>":"";y+=O;L--;D.length--}var r=D[D.length-1];D.length--;D[D.length-1].buffer="";if(r.starts){I(r.starts,"")}return R.rE}if(w(M,R)){throw"Illegal"}}var E=e[B];var D=[E.dM];var A=0;var x=0;var y="";try{var s,u=0;E.dM.buffer="";do{s=p(C,u);var t=G(s[0],s[1],s[2]);u+=s[0].length;if(!t){u+=s[1].length}}while(!s[2]);if(D.length>1){throw"Illegal"}return{r:A,keyword_count:x,value:y}}catch(H){if(H=="Illegal"){return{r:0,keyword_count:0,value:m(C)}}else{throw H}}}function g(t){var p={keyword_count:0,r:0,value:m(t)};var r=p;for(var q in e){if(!e.hasOwnProperty(q)){continue}var s=d(q,t);s.language=q;if(s.keyword_count+s.r>r.keyword_count+r.r){r=s}if(s.keyword_count+s.r>p.keyword_count+p.r){r=p;p=s}}if(r.language){p.second_best=r}return p}function i(r,q,p){if(q){r=r.replace(/^((<[^>]+>|\t)+)/gm,function(t,w,v,u){return w.replace(/\t/g,q)})}if(p){r=r.replace(/\n/g,"<br>")}return r}function n(t,w,r){var x=h(t,r);var v=a(t);var y,s;if(v){y=d(v,x)}else{return}var q=c(t);if(q.length){s=document.createElement("pre");s.innerHTML=y.value;y.value=k(q,c(s),x)}y.value=i(y.value,w,r);var u=t.className;if(!u.match("(\\s|^)(language-)?"+v+"(\\s|$)")){u=u?(u+" "+v):v}if(/MSIE [678]/.test(navigator.userAgent)&&t.tagName=="CODE"&&t.parentNode.tagName=="PRE"){s=t.parentNode;var p=document.createElement("div");p.innerHTML="<pre><code>"+y.value+"</code></pre>";t=p.firstChild.firstChild;p.firstChild.cN=s.cN;s.parentNode.replaceChild(p.firstChild,s)}else{t.innerHTML=y.value}t.className=u;t.result={language:v,kw:y.keyword_count,re:y.r};if(y.second_best){t.second_best={language:y.second_best.language,kw:y.second_best.keyword_count,re:y.second_best.r}}}function o(){if(o.called){return}o.called=true;var r=document.getElementsByTagName("pre");for(var p=0;p<r.length;p++){var q=b(r[p]);if(q){n(q,hljs.tabReplace)}}}function l(){if(window.addEventListener){window.addEventListener("DOMContentLoaded",o,false);window.addEventListener("load",o,false)}else{if(window.attachEvent){window.attachEvent("onload",o)}else{window.onload=o}}}var e={};this.LANGUAGES=e;this.highlight=d;this.highlightAuto=g;this.fixMarkup=i;this.highlightBlock=n;this.initHighlighting=o;this.initHighlightingOnLoad=l;this.IR="[a-zA-Z][a-zA-Z0-9_]*";this.UIR="[a-zA-Z_][a-zA-Z0-9_]*";this.NR="\\b\\d+(\\.\\d+)?";this.CNR="\\b(0[xX][a-fA-F0-9]+|(\\d+(\\.\\d*)?|\\.\\d+)([eE][-+]?\\d+)?)";this.BNR="\\b(0b[01]+)";this.RSR="!|!=|!==|%|%=|&|&&|&=|\\*|\\*=|\\+|\\+=|,|\\.|-|-=|/|/=|:|;|<|<<|<<=|<=|=|==|===|>|>=|>>|>>=|>>>|>>>=|\\?|\\[|\\{|\\(|\\^|\\^=|\\||\\|=|\\|\\||~";this.ER="(?![\\s\\S])";this.BE={b:"\\\\.",r:0};this.ASM={cN:"string",b:"'",e:"'",i:"\\n",c:[this.BE],r:0};this.QSM={cN:"string",b:'"',e:'"',i:"\\n",c:[this.BE],r:0};this.CLCM={cN:"comment",b:"//",e:"$"};this.CBLCLM={cN:"comment",b:"/\\*",e:"\\*/"};this.HCM={cN:"comment",b:"#",e:"$"};this.NM={cN:"number",b:this.NR,r:0};this.CNM={cN:"number",b:this.CNR,r:0};this.BNM={cN:"number",b:this.BNR,r:0};this.inherit=function(r,s){var p={};for(var q in r){p[q]=r[q]}if(s){for(var q in s){p[q]=s[q]}}return p}}();hljs.LANGUAGES.cpp=function(){var a={keyword:{"false":1,"int":1,"float":1,"while":1,"private":1,"char":1,"catch":1,"export":1,virtual:1,operator:2,sizeof:2,dynamic_cast:2,typedef:2,const_cast:2,"const":1,struct:1,"for":1,static_cast:2,union:1,namespace:1,unsigned:1,"long":1,"throw":1,"volatile":2,"static":1,"protected":1,bool:1,template:1,mutable:1,"if":1,"public":1,friend:2,"do":1,"return":1,"goto":1,auto:1,"void":2,"enum":1,"else":1,"break":1,"new":1,extern:1,using:1,"true":1,"class":1,asm:1,"case":1,typeid:1,"short":1,reinterpret_cast:2,"default":1,"double":1,register:1,explicit:1,signed:1,typename:1,"try":1,"this":1,"switch":1,"continue":1,wchar_t:1,inline:1,"delete":1,alignof:1,char16_t:1,char32_t:1,constexpr:1,decltype:1,noexcept:1,nullptr:1,static_assert:1,thread_local:1,restrict:1,_Bool:1,complex:1},built_in:{std:1,string:1,cin:1,cout:1,cerr:1,clog:1,stringstream:1,istringstream:1,ostringstream:1,auto_ptr:1,deque:1,list:1,queue:1,stack:1,vector:1,map:1,set:1,bitset:1,multiset:1,multimap:1,unordered_set:1,unordered_map:1,unordered_multiset:1,unordered_multimap:1,array:1,shared_ptr:1}};return{dM:{k:a,i:"</",c:[hljs.CLCM,hljs.CBLCLM,hljs.QSM,{cN:"string",b:"'\\\\?.",e:"'",i:"."},{cN:"number",b:"\\b(\\d+(\\.\\d*)?|\\.\\d+)(u|U|l|L|ul|UL|f|F)"},hljs.CNM,{cN:"preprocessor",b:"#",e:"$"},{cN:"stl_container",b:"\\b(deque|list|queue|stack|vector|map|set|bitset|multiset|multimap|unordered_map|unordered_set|unordered_multiset|unordered_multimap|array)\\s*<",e:">",k:a,r:10,c:["self"]}]}}}();hljs.LANGUAGES.r={dM:{c:[hljs.HCM,{cN:"number",b:"\\b0[xX][0-9a-fA-F]+[Li]?\\b",e:hljs.IMMEDIATE_RE,r:0},{cN:"number",b:"\\b\\d+(?:[eE][+\\-]?\\d*)?L\\b",e:hljs.IMMEDIATE_RE,r:0},{cN:"number",b:"\\b\\d+\\.(?!\\d)(?:i\\b)?",e:hljs.IMMEDIATE_RE,r:1},{cN:"number",b:"\\b\\d+(?:\\.\\d*)?(?:[eE][+\\-]?\\d*)?i?\\b",e:hljs.IMMEDIATE_RE,r:0},{cN:"number",b:"\\.\\d+(?:[eE][+\\-]?\\d*)?i?\\b",e:hljs.IMMEDIATE_RE,r:1},{cN:"keyword",b:"(?:tryCatch|library|setGeneric|setGroupGeneric)\\b",e:hljs.IMMEDIATE_RE,r:10},{cN:"keyword",b:"\\.\\.\\.",e:hljs.IMMEDIATE_RE,r:10},{cN:"keyword",b:"\\.\\.\\d+(?![\\w.])",e:hljs.IMMEDIATE_RE,r:10},{cN:"keyword",b:"\\b(?:function)",e:hljs.IMMEDIATE_RE,r:2},{cN:"keyword",b:"(?:if|in|break|next|repeat|else|for|return|switch|while|try|stop|warning|require|attach|detach|source|setMethod|setClass)\\b",e:hljs.IMMEDIATE_RE,r:1},{cN:"literal",b:"(?:NA|NA_integer_|NA_real_|NA_character_|NA_complex_)\\b",e:hljs.IMMEDIATE_RE,r:10},{cN:"literal",b:"(?:NULL|TRUE|FALSE|T|F|Inf|NaN)\\b",e:hljs.IMMEDIATE_RE,r:1},{cN:"identifier",b:"[a-zA-Z.][a-zA-Z0-9._]*\\b",e:hljs.IMMEDIATE_RE,r:0},{cN:"operator",b:"<\\-(?!\\s*\\d)",e:hljs.IMMEDIATE_RE,r:2},{cN:"operator",b:"\\->|<\\-",e:hljs.IMMEDIATE_RE,r:1},{cN:"operator",b:"%%|~",e:hljs.IMMEDIATE_RE},{cN:"operator",b:">=|<=|==|!=|\\|\\||&&|=|\\+|\\-|\\*|/|\\^|>|<|!|&|\\||\\$|:",e:hljs.IMMEDIATE_RE,r:0},{cN:"operator",b:"%",e:"%",i:"\\n",r:1},{cN:"identifier",b:"`",e:"`",r:0},{cN:"string",b:'"',e:'"',c:[hljs.BE],r:0},{cN:"string",b:"'",e:"'",c:[hljs.BE],r:0},{cN:"paren",b:"[[({\\])}]",e:hljs.IMMEDIATE_RE,r:0}]}};
hljs.initHighlightingOnLoad();
</script>




</head>

<body>
<h1>Using CIMPLR to call Common Insertion Sites using bias maps without permutation testing</h1>

<p>In preparation for running the actual CIMPLR commands, we start by preparing an R environment containing CIMPLR functions and insertion data.</p>

<p>First set your working directory to the root folder of the package. Here we do that by setting the &#39;Knit&#39; options to create this Vignette; use <code>setwd(&#39;../&#39;)</code> at the console.</p>

<pre><code class="r">opts_knit$set(root.dir = &quot;../&quot;)
</code></pre>

<p>Load the necessary libraries and sources of the package.</p>

<pre><code class="r">
library(Biobase)
</code></pre>

<pre><code>## Loading required package: BiocGenerics
## Loading required package: parallel
## 
## Attaching package: &#39;BiocGenerics&#39;
## 
## The following objects are masked from &#39;package:parallel&#39;:
## 
##     clusterApply, clusterApplyLB, clusterCall, clusterEvalQ,
##     clusterExport, clusterMap, parApply, parCapply, parLapply,
##     parLapplyLB, parRapply, parSapply, parSapplyLB
## 
## The following object is masked from &#39;package:stats&#39;:
## 
##     xtabs
## 
## The following objects are masked from &#39;package:base&#39;:
## 
##     Filter, Find, Map, Position, Reduce, anyDuplicated,
##     as.data.frame, cbind, colnames, duplicated, eval, get,
##     intersect, lapply, mapply, match, mget, order, paste, pmax,
##     pmax.int, pmin, pmin.int, rank, rbind, rep.int, rownames,
##     sapply, setdiff, sort, table, tapply, union, unique, unlist
## 
## Welcome to Bioconductor
## 
##     Vignettes contain introductory material; view with
##     &#39;browseVignettes()&#39;. To cite Bioconductor, see
##     &#39;citation(&quot;Biobase&quot;)&#39;, and for packages &#39;citation(&quot;pkgname&quot;)&#39;.
</code></pre>

<pre><code class="r">library(GenomicFeatures)
</code></pre>

<pre><code>## Loading required package: IRanges
## Loading required package: GenomicRanges
## Loading required package: AnnotationDbi
</code></pre>

<pre><code class="r">library(rjson)
library(multicore)
</code></pre>

<pre><code>## 
## Attaching package: &#39;multicore&#39;
## 
## The following objects are masked from &#39;package:parallel&#39;:
## 
##     mclapply, mcparallel, pvec
</code></pre>

<pre><code class="r">library(rtracklayer)
library(raster)
</code></pre>

<pre><code>## Loading required package: sp
## 
## Attaching package: &#39;sp&#39;
## 
## The following object is masked from &#39;package:IRanges&#39;:
## 
##     %over%
## 
## 
## Attaching package: &#39;raster&#39;
## 
## The following object is masked from &#39;package:rtracklayer&#39;:
## 
##     values
## 
## The following objects are masked from &#39;package:AnnotationDbi&#39;:
## 
##     direction, select
## 
## The following objects are masked from &#39;package:GenomicRanges&#39;:
## 
##     distance, shift, values, values&lt;-
## 
## The following objects are masked from &#39;package:IRanges&#39;:
## 
##     distance, shift, trim, values, values&lt;-
</code></pre>

<pre><code class="r">
source(&quot;R/AllClasses.R&quot;)
source(&quot;R/AllGenerics.R&quot;)
</code></pre>

<pre><code>## Creating a new generic function for &#39;metadata&#39; in the global environment
</code></pre>

<pre><code class="r">source(&quot;R/KSEDistribution.R&quot;)
source(&quot;R/core.R&quot;)
source(&quot;R/cimplr.R&quot;)
</code></pre>

<p>Load your data as a <code>data.frame</code> containing at least a <code>chr</code> and <code>location</code> column.</p>

<pre><code class="r">load(&quot;data/test_iset.rda&quot;)

head(iset)
</code></pre>

<pre><code>##     chr location
## 1 chr10 14550904
## 2 chr10 34266702
## 3 chr10 45748052
## 4 chr10 35169341
## 5 chr10 19266925
## 6 chr10 52070550
</code></pre>

<p>Now we can start with CIMPLR.</p>

<h2>Step 1: Initialise a CIMPLR object</h2>

<p>We construct a CIMPLR object based on the insertion data and the parameters for the CIMPLR run.</p>

<pre><code class="r">co &lt;- cimplr(insertions = iset, reference = &quot;mm9&quot;, pattern = &quot;TA&quot;, scales = log.seq(5000, 
    5e+05, 100)[seq(1, 50, 5)], pkse.method = &quot;count&quot;, alpha.level = 0.05, p.adjust.method = &quot;bonferroni&quot;, 
    p.adjust.n.method = &quot;n.x&quot;, mtf = 1, test.chromosomes.seperately = TRUE, 
    n.cores = c(scales = 10, chromosomes = 1), genes.file = NULL)
</code></pre>

<pre><code>## *** CIMPLR *** 
## 
## List of 22
##  $ n.bp.total                 : int 129993255
##  $ n.bgsites.total            : int 8228738
##  $ n.insertions.total         : int 1000
##  $ chr.info                   :&#39;data.frame&#39;: 1 obs. of  3 variables:
##   ..$ length      : int 129993255
##   ..$ TA.count    : int 8228738
##   ..$ n.insertions: int 1000
##  $ insertions.org             :&#39;data.frame&#39;: 1000 obs. of  2 variables:
##   ..$ chr     : chr [1:1000] &quot;chr10&quot; &quot;chr10&quot; &quot;chr10&quot; &quot;chr10&quot; ...
##   ..$ location: num [1:1000] 14550904 34266702 45748052 35169341 19266925 ...
##  $ chromosomes                : chr &quot;chr10&quot;
##  $ insertions                 :List of 1
##   ..$ chr10: num [1:1000] 14550904 34266702 45748052 35169341 19266925 ...
##  $ reference                  : chr &quot;mm9&quot;
##  $ pattern                    : chr &quot;TA&quot;
##  $ exclude.chromosomes        : NULL
##  $ scales                     : num [1:10] 5000 6309 7961 10046 12677 ...
##  $ pkse.method                : chr &quot;count&quot;
##  $ alpha.level                : num 0.05
##  $ p.adjust.method            : chr &quot;bonferroni&quot;
##  $ p.adjust.n.method          : chr &quot;n.x&quot;
##  $ mtf                        : num 1
##  $ test.chromosomes.seperately: logi TRUE
##  $ stepsize                   : num 1000
##  $ D                          : num 4
##  $ genes.file                 : NULL
##  $ reference.dir              : chr &quot;./references&quot;
##  $ n.cores                    : Named num [1:2] 10 1
##   ..- attr(*, &quot;names&quot;)= chr [1:2] &quot;scales&quot; &quot;chromosomes&quot;
</code></pre>

<h2>Step 2: Load a bias map</h2>

<p>As a bias map, we load the TA count files under the &#39;./references&#39; folder. This bias map is simply the TA site background, but more sophisticated bias maps can be supplied here (link to CreateBiasMap Vignette).</p>

<pre><code class="r">co &lt;- cimplr.reference(co)
</code></pre>

<pre><code>## cimplr.reference()
</code></pre>

<h2>Step 3: Gaussian Kernel Convolution</h2>

<p>With the insertion data and the bias map loaded and pre-processed within the CIMPLR object, we can now start the kernel convolution, resulting in Kernel Smoothed Estimates (KSEs) for all scales.</p>

<pre><code class="r">co &lt;- cimplr.convolve(co)
</code></pre>

<pre><code>## cimplr.convolve()
</code></pre>

<h2>Step 4: Calculate the threshold</h2>

<p>A seperate step calculates the threshold for the KSE at each scale. This is done using the analytical KSE distribution at a controlled FDR and taking into account multiple scales (link to KSEdistribution Vignette).</p>

<pre><code class="r">co &lt;- cimplr.threshold(co)
</code></pre>

<pre><code>## cimplr.threshold()
## calc.fdr.pval()
</code></pre>

<h2>Step 5: Call CISes</h2>

<p>With the thresholds in place, the last step is to do the CIS calling. This returns two lists of CISes: (1) all CIS for each scale, and (2) the cross scale CIS using the &#39;cis-cloud-global-maximum&#39; procedure (link to CrossScaleCIScalling Vignette)</p>

<pre><code class="r">co &lt;- cimplr.call(co)
</code></pre>

<pre><code>## cimplr.call()
## cimplr.call() - merge CIS list
## cimplr.call() - call CIS across scales (chr = chr10)
</code></pre>

<pre><code class="r">

head(co$output$all.cises)
</code></pre>

<pre><code>##                     chromosome    start      end width peak.location
## CIS10:5941001_5000       chr10  5919001  5941001 22000       5941001
## CIS10:11623001_5000      chr10 11620001 11623001  3000      11623001
## CIS10:57203001_5000      chr10 57203001 57234001 31000      57203001
## CIS10:71380001_5000      chr10 71380001 71388001  8000      71380001
## CIS10:72966001_5000      chr10 72959001 72966001  7000      72966001
## CIS10:76959001_5000      chr10 76959001 76978001 19000      76959001
##                     peak.kse.value peak.bg peak.ratio
## CIS10:5941001_5000           2.317    2262      1.149
## CIS10:11623001_5000          2.319    2739      1.041
## CIS10:57203001_5000          2.312    2905      1.008
## CIS10:71380001_5000          2.238    2722      1.009
## CIS10:72966001_5000          2.350    3032      1.004
## CIS10:76959001_5000          1.958    1625      1.003
##                     n.insertions.within.3.sigma scale alpha.level
## CIS10:5941001_5000                           13  5000        0.05
## CIS10:11623001_5000                           3  5000        0.05
## CIS10:57203001_5000                          12  5000        0.05
## CIS10:71380001_5000                           4  5000        0.05
## CIS10:72966001_5000                           4  5000        0.05
## CIS10:76959001_5000                           5  5000        0.05
##                     p.adjust.n p.adjust.method p.adjust.n.method mtf
## CIS10:5941001_5000      129994      bonferroni               n.x   1
## CIS10:11623001_5000     129994      bonferroni               n.x   1
## CIS10:57203001_5000     129994      bonferroni               n.x   1
## CIS10:71380001_5000     129994      bonferroni               n.x   1
## CIS10:72966001_5000     129994      bonferroni               n.x   1
## CIS10:76959001_5000     129994      bonferroni               n.x   1
##                     adjusted.alpha.level
## CIS10:5941001_5000             3.846e-07
## CIS10:11623001_5000            3.846e-07
## CIS10:57203001_5000            3.846e-07
## CIS10:71380001_5000            3.846e-07
## CIS10:72966001_5000            3.846e-07
## CIS10:76959001_5000            3.846e-07
</code></pre>

<pre><code class="r">head(co$output$collapsed.cises)
</code></pre>

<pre><code>##                      chromosome    start      end width peak.location
## CIS10:5941001_5000        chr10  5919001  5941001 22000       5941001
## CIS10:11623001_5000       chr10 11620001 11623001  3000      11623001
## CIS10:57243001_12676      chr10 57196001 57243001 47000      57243001
## CIS10:71379001_10046      chr10 71379001 71392001 13000      71379001
## CIS10:72957001_10046      chr10 72957001 72969001 12000      72957001
## CIS10:76498001_6309       chr10 76498001 76500001  2000      76498001
##                      peak.kse.value peak.bg peak.ratio
## CIS10:5941001_5000            2.317    2262      1.149
## CIS10:11623001_5000           2.319    2739      1.041
## CIS10:57243001_12676          3.339    7757      1.031
## CIS10:71379001_10046          2.967    5605      1.031
## CIS10:72957001_10046          3.058    6097      1.041
## CIS10:76498001_6309           2.030    2198      1.016
##                      n.insertions.within.3.sigma scale alpha.level
## CIS10:5941001_5000                            13  5000        0.05
## CIS10:11623001_5000                            3  5000        0.05
## CIS10:57243001_12676                          14 12677        0.05
## CIS10:71379001_10046                           5 10046        0.05
## CIS10:72957001_10046                           4 10046        0.05
## CIS10:76498001_6309                            3  6309        0.05
##                      p.adjust.n p.adjust.method p.adjust.n.method mtf
## CIS10:5941001_5000       129994      bonferroni               n.x   1
## CIS10:11623001_5000      129994      bonferroni               n.x   1
## CIS10:57243001_12676     129994      bonferroni               n.x   1
## CIS10:71379001_10046     129994      bonferroni               n.x   1
## CIS10:72957001_10046     129994      bonferroni               n.x   1
## CIS10:76498001_6309      129994      bonferroni               n.x   1
##                      adjusted.alpha.level
## CIS10:5941001_5000              3.846e-07
## CIS10:11623001_5000             3.846e-07
## CIS10:57243001_12676            3.846e-07
## CIS10:71379001_10046            3.846e-07
## CIS10:72957001_10046            3.846e-07
## CIS10:76498001_6309             3.846e-07
</code></pre>

</body>

</html>

